### DISCUSSION TITLE
Submarine hull revamp

### DISCUSSION THREAD

**Discussion by Glandziu** [FakeFishGames/Barotrauma#9599] [Ideas and suggestions]
Rectangular hulls are a pain to work with if your submarine is supposed to have more funky looks on a bigger scale. A simple curve or even a wall at an angle can add a fair few hulls and even more gaps if it's steep enough and f.e. runs between two rooms on the sub. This wouldn't be much of a problem if not for the janky water behaviour that smaller hulls create - as reported in [ #9279](https://github.com/Regalis11/Barotrauma/issues/9279) - where there can be large spikes of water throwing the crew about before it eventually equalises with others as the room fills up (or empties) completely. This is of course omitting the pain the submarine builder has to go through to hull and gap everything properly.

However, this behaviour can be removed for the most of regular usage of small, connected with gaps hulls, additionally greatly expanding on the capabilities of the editor and making the submarine building process more user-friendly and maybe even improving performance, hence the idea:

**Hulls that are interconnected to form a single room would be calculated as one, bigger hull.**

Calculating the fill percentages of water height from the bottom to the top of the connected hulls could for example be solved by:

> In editor - measuring the horizontal cross-section of a linked hull, calculating the volume at which the cross-section changes, when measured from the bottom and saving it as a simple array connected to the hull and containing:
> - a volume value of the amount of hull volume below - this is the horizontal marker,
> - a cross-section of the section above the horizontal marker (until the next marker this is static),
> - the height of hull from the lowest point to the marker
> 
> The in-game code could be based on water ingress/egress in hull units, the current level of water in hull units and the aforementioned array to determine the correct water height with each update. It could be implemented as this:
> - Adding the amount of water going in or going out of the hull to get the current water level.
> - Setting the fill level height to the value related to the closest, smaller-than volume amount specified in the array when comparing with the amount of water in hull units, then subtracting that volume from the current water level in a local variable.
> - The rest of the height would be added in based on the cross-section value from the array and remaining amount of water.
>
> Concerning the graphical side of things, the water could be masked outside of the linked hull containing it and still use simple height rectangular fill methods for the most of it.
> Likewise, the physics calculations could be run on the already masked water (and if not possible with a simple check of being inside the hull) thus not affecting the outside of a linked hull.
> 
> This example is still lightweight resource-wise, relatively simple to implement and should keep the current way of hulling still working as legacy if auto-updating the hulls as single entry arrays.
>Additionally, with some clever math it could even allow to be further modified to accept any kind of geometric shape ([maybe based on polygonal segments](https://docs.gimp.org/2.10/en/gimp-tool-free-select.html)) ;)

TL;DR: Treating multiple hulls linked into one room as one big hull might likely get rid of some if not all of the gaps between rooms that in reality shouldn't have any, hopefully leading to better performance.

**Comment by MaaiG** [comment]
I just add it for quantity https://github.com/Regalis11/Barotrauma/discussions/6172
PS And don’t forget the lights.
